### PR TITLE
Add ability to sign using taproot tweaks

### PR DESCRIPTION
**Pull request by xoloki**
Taproot defines a key tweaking algorithm that allow either direct key spends, or script spends. The scripts are contained in a tree of scripts, with a merkle tree (MAST) to represent them. The tweaking algorithm adds a hash to the real public and private keys, which is needed for the direct key spends; the script spends use the real public key combined with the merkle root to create a valid schnorr signature.

If you only want to allow key spends, then you could avoid adding the tweak, and just use the real public key to form a standard (or aggregated) schnorr proof. This works fine in the non-aggregated case. But in the aggregated case, it is possible for a malicious participant in the distributed key generation protocol to massage their part of the key to include a hidden script spend. This requires that the malicious participant go last, but otherwise is trivial to execute.

To avoid this, we must always tweak the public key which is used in a taproot output. This tweak consists of a hash of the real public key and the root of the MAST. Since we are adding this tweak to the public key, we must also add it to the private key. But we don't actually recover the private key explicitly when computing an aggregate signature; however, if we multiply the tweak by the challenge then add to the signature, then the math works out such that we have effectively added the tweak to the private key.

### CODE REVIEW

File: src/compute.rs [hunk 1]
```diff
@@ -131,3 +125,41 @@ pub fn poly(x: &Scalar, f: &Vec<Point>) -> Result<Point, PointError> {
...
+    tag_hasher.update(tag.as_bytes());
+    let tag_hash = tag_hasher.finalize();
+
+    hasher.update(tag_hash);
+    hasher.update(tag_hash);
+
+    hasher
+}
+
+/// Create a BIP341 compliant taproot tweak from a public key and merkle root
```
Comment (netrome) [inline src/compute.rs:143]: I saw you changed this, but isn't BIP340 (Schnorr signatures) the one that defines the details of the tweak - while it's used in BIP341 (Taproot)?

File: src/taproot.rs [hunk 2]
```diff
@@ -168,13 +177,42 @@ pub mod test_helpers {
...
+            signers.iter_mut().flat_map(|s| s.gen_nonces(rng)).collect();
+
+        loop {
+            let (_, R) = Signer::compute_intermediate(msg, &signer_ids, &key_ids, &nonces);
+            if R.has_even_y() {
+                break;
+            }
+            nonces = signers.iter_mut().flat_map(|s| s.gen_nonces(rng)).collect();
+        }
+
```
Comment (netrome) [inline src/taproot.rs]: Nit: From what I can see, this is the same code as in the `.sign(..)` method and could be factored out to a shared helper function for increased code reuse and readability.

**Review by netrome — APPROVED** [review]
Nice stuff. I'm surprised by how few changes you had to introduce support for the tap tweak.

File: src/compute.rs (hunk 1 above)
Comment (xoloki) [reply src/compute.rs:143]: Exactly, BIP340 defines Schnorr, while BIP341 defines taproot, and BIP342 defines tapscript.

So `compute::{challenge, tagged_hash}` are BIP340, while `compute::{tweak, tweaked_public_key, merkle_root}` are BIP341.

File: src/taproot.rs (hunk 2 above)
Comment (xoloki) [reply src/taproot.rs]: Yeah, I did a better job factoring `SignatureAggregator::{sign, sign_merkle_root, sign_with_tweak}` for code reuse.  I'll do something similar for `test_helpers::{sign, sign_tweaked}`.

Comment (xoloki) [reply src/taproot.rs]: Done.

**Comment by xoloki** [comment]
> Nice stuff. I'm surprised by how few changes you had to introduce support for the tap tweak.

It was reasonably straightforward once I actually understood BIP341.  I almost got away with putting all of the logic into `SignatureAggregator` but unfortunately the signers needed to use the tweaked public key to compute the challenge.

I think I can do away with the multiple DKG rounds in a later PR, but I wanted to get it working without that since the math gets a little ugly to deal with odd public keys.  As I keep saying to @jferrant it's always better to start with something ugly that works :P

**Comment by xoloki** [comment]
Ok, I've got `v2` working now with a taproot test, will squamerge (squerge?) once it goes green.  Thanks for the review @netrome !

**Pull request merged**